### PR TITLE
Implement LG language features in VS Code extension with Language Server Protocol.

### PR DESCRIPTION
**Pull request by ShYuPe**
# Implement LG extension using Language Server Protocol.

closes: #2566 

We implemented LG extension using Language Server Protocol to unify the experience of VS Code extension and Composer and improve the performance. 

The extensions of LG we implemented using LSP includes syntax highlighting, diagnostic check, template reference hover, buildin function hover, buildin function and template suggestion, structure property suggestion and template definition.

### CODE REVIEW

**Comment by boydc2014** [comment]
@cosmicshuai can you add more context of why we want to switch to LSP for better transparency to other reviewers.

File: experimental/BotBuilder-lsp-lg/server/src/server.ts [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+			workspaceFolders = workspaceFolders?.filter(workspaceFolder => !_event.removed.includes(workspaceFolder));
+			_event.added.forEach(folderAdded => workspaceFolders?.push(folderAdded));
+			util.triggerLGFileFinder(workspaceFolders!);
+			connection.console.log('Workspace folder change event received.');
+		});
+	}
+});
+
+// The example settings
+interface ExampleSettings {
```
Comment (cosmicshuai) [inline experimental/BotBuilder-lsp-lg/server/src/server.ts:133]: change the naming of ExampleSettings. You can use LGSettings

File: experimental/BotBuilder-lsp-lg/client/src/extension.ts [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+				didChangeWorkspaceFolders: ((data, next) => {
+					next(data);
+				})
+			}
+		},
+		diagnosticCollectionName: 'lg'
+	};
+
+	// Create the language client and start the client.
+	client = new LanguageClient(
```
Comment (cosmicshuai) [inline experimental/BotBuilder-lsp-lg/client/src/extension.ts:66]: rename the client

**Comment by cosmicshuai** [comment]
The reason we switch our extension to LSP is we want to unify the experience of VS Code extension and Composer. Making the implementation through LSP will reduce the work of adding new language features. Moreover using LSP will help improve the performance.

File: experimental/BotBuilder-lsp-lg/server/src/providers/completion.ts [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+		const paths = Array.from(new Set(TemplatesStatus.lgFilesOfWorkspace));
+
+		return paths.filter(u => fspath !== u).reduce((prev : any[], curr: string) => {
+			const relativePath = path.relative(path.dirname(fspath!), curr);
+			const item = {
+				label: relativePath, 
+				kind: CompletionItemKind.Reference,
+				detail: curr
+			};
+			item.detail = curr;
```
[the remark this answers is not in the document]
Comment (ShYuPe) [reply experimental/BotBuilder-lsp-lg/server/src/providers/completion.ts]: fixed

File: experimental/BotBuilder-lsp-lg/server/src/providers/completion.ts [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+			});
+			return items;
+		}
+	}  else if (/^>\s!#/.test(lineTextBefore!)) {
+		// options suggestion following "> !#"
+		const items: CompletionItem[] = [];
+		if (/>\s!#$/.test(lineTextBefore!)) {
+			AddToCompletion(util.optionsMap.options, items);
+		} else {
+			if (/>\s!#\s*@strict\s*=$/.test(lineTextBefore!)) {
```
[the remark this answers is not in the document]
Comment (ShYuPe) [reply experimental/BotBuilder-lsp-lg/server/src/providers/completion.ts]: fixed

File: experimental/BotBuilder-lsp-lg/server/src/providers/completion.ts [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+
+function isStartStructure(document: TextDocument,
+	position: Position) {
+		const lineTextBefore = document?.getText({
+			start: {
+				line: position.line,
+				character: 0
+			},
+			end: position
+		}).toString();
```
[the remark this answers is not in the document]
Comment (ShYuPe) [reply experimental/BotBuilder-lsp-lg/server/src/providers/completion.ts]: fixed

File: experimental/BotBuilder-lsp-lg/server/src/server.ts (hunk 1 above)
Comment (ShYuPe) [reply experimental/BotBuilder-lsp-lg/server/src/server.ts]: fixed

File: experimental/BotBuilder-lsp-lg/client/src/extension.ts (hunk 2 above)
Comment (ShYuPe) [reply experimental/BotBuilder-lsp-lg/client/src/extension.ts]: fixed

**Comment by Danieladu** [comment]
Please add the generated files into .gitignore.

**Comment by Danieladu** [comment]
Add more instructions or guidance including but not limited to the usage/launch file details/how to contribute/LG lsp introduction .etc.

File: experimental/adaptive-tool/client/tsconfig.json [hunk 6]
```diff
@@ -0,0 +1,12 @@
+{
+	"compilerOptions": {
+		"module": "commonjs",
+		"target": "es2019",
```
Comment (Danieladu) [inline experimental/adaptive-tool/client/tsconfig.json:4]: it is better to keep the "lib" and "target" be consistent with other typescript projects.

**Comment by Danieladu** [comment]
At present, many providers of vscode are relatively concise. You can refer to the lgserver in Composer to complete some provider functions. refer: https://github.com/microsoft/BotFramework-Composer/blob/d9bd8968fede55b581066b241522dc862b4aba7a/Composer/packages/tools/language-servers/language-generation/src/LGServer.ts#L480

For Example, vscode uses the simple regular matches to achieve the Completion feature. But  Composer uses FSM (Finite State Machine) to traverse all the states. You can compare the pros and cons of these two solutions, and find what corner cases are not covered by the simple regular matching. 

The process of exploration is also the process of learning.

File: experimental/adaptive-tool/server/lg/src/providers/completion.ts [hunk 7]
```diff
@@ -0,0 +1,211 @@
...
+} from 'vscode-languageserver-textdocument';
+import * as util from '../util';
+import { TemplatesStatus } from '../templatesStatus';
+import * as path from 'path';
+
+export function provideCompletionItems(_textDocumentPosition: TextDocumentPositionParams, documents: TextDocuments<TextDocument>){
+	const document = documents.get(_textDocumentPosition.textDocument.uri);
+	const fspath = Files.uriToFilePath(document!.uri);
+	const position = _textDocumentPosition.position;
+	const lineTextBefore = document?.getText({
```
Comment (Danieladu) [inline experimental/adaptive-tool/server/lg/src/providers/completion.ts:21]: As I know, [Optional Chaining](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-7.html#optional-chaining) first appeared on ts 3.7 version. So, you should make sure all the compile tools (dependency, typescript version of vscode) should be newer than version 3.7.

File: experimental/adaptive-tool/server/lg/src/util.ts [hunk 8]
```diff
@@ -0,0 +1,327 @@
+
+import { TextDocument, Range, Position } from "vscode-languageserver-textdocument";
```
Comment (Danieladu) [inline experimental/adaptive-tool/server/lg/src/util.ts:2]: copyright

Comment (ShYuPe) [reply experimental/adaptive-tool/server/lg/src/util.ts:6]: fixed

File: experimental/adaptive-tool/client/tsconfig.json (hunk 6 above)
Comment (ShYuPe) [reply experimental/adaptive-tool/client/tsconfig.json]: fixed

File: experimental/adaptive-tool/server/lg/src/providers/completion.ts (hunk 7 above)
Comment (ShYuPe) [reply experimental/adaptive-tool/server/lg/src/providers/completion.ts:21]: fixed

**Comment by ShYuPe** [comment]
> Please create a detailed github issue describing what this is, and the requirements around it. That will give us enough context to review.

Issue#2566 has been linked with this pr

**Comment by cosmicshuai** [comment]
I installed the generated vsix file into my VS Code client. Tested several .lg file. LGTM.

**Comment by ShYuPe** [comment]
@cleemullins 
We create a detailed github issue to describe this pr.

**Comment by boydc2014** [comment]
@cleemullins this one looks good to me now, this is a converge of previous different impelmentation across vscode extension and composer(manaco) extension.

**Comment by boydc2014** [comment]
@cleemullins is the issue description looks good to you now? we want to take this and next we are migrating the lu implementation into lsp also, then we will have a full converged implementation that shared across composer and vscode extension.

**Pull request merged**